### ISSUE TITLE
[Proposal] Builder::each() should have a dedicated implementation

### ISSUE THREAD

**Issue by QuakePhil** [laravel/internals#882]
According to https://github.com/illuminate/database/blob/master/Concerns/BuildsQueries.php#L58 and https://github.com/laravel/framework/issues/14007 `each()` currently leverages `chunk()` by design.  Ran into a problem where `chunk()` was slowing down significantly as the "OFFSET" part of the query became sufficiently large.  (Wrote a faster custom chunker by exploiting some assumptions about data and doing a whereIn() on PK) However, I still need to use `each()` elsewhere...

My current workaround is something like this (but it depends on getPdo() and perhaps $row should be `collect()`ed, ideally a better version would use some sort of native `next()` api which I could not find)

```
    /**
     * Execute a callback over each item.
     *
     * @param  callable  $callback
     * @return bool
     */
    public function each(callable $callback)
    {
        $results = app('db')->getPdo()->query($this->toSql());
        if ($results === false) {
            return false;
        }

        while ($row = $results->fetch(\PDO::FETCH_ASSOC)) {
            if ($callback($row) === false) {
                return false;
            }
        }
        return true;
    }
```

So, to recap, even if I use native `each(function($item){...}, 1)` I still end up with `LIMIT 1 OFFSET 1293871823714` and that takes a while.  Can we re-write (may require a new next() api?) each() in a more straightforward way where it is iterating a result (as some might expect `each` to do)?

By way of https://github.com/laravel/framework/issues/22061 a little more detail there

**Comment by mfn** [comment]
But your suggestion would still load the whole dataset into memory, wouldn't it?

Depending on the size of your date, this may be even a worse solution then the dreaded offset-based pagination.

**Comment by QuakePhil** [comment]
> But your suggestion would still load the whole dataset into memory, wouldn't it?

@mfn I don't see how it would...  Perhaps you can reference the line number where the whole dataset is loaded into memory in the above code snippet?

**Comment by sisve** [comment]
Mysql (or at least the pdo_mysql connector) defaults to buffered queries, which reads the entire result into the memory. You need to set `PDO::MYSQL_ATTR_USE_BUFFERED_QUERY` to false for this to stream the query result, but that won't work with other databases. I'm not sure how the other pdo connectors behave.

Related: https://dev.mysql.com/doc/apis-php/en/apis-php-mysqlinfo.concepts.buffering.html

**Comment by QuakePhil** [comment]
> ... You need to set `PDO::MYSQL_ATTR_USE_BUFFERED_QUERY` to false for this to stream the query result, but that won't work with other databases. I'm not sure how the other pdo connectors behave.

That is correct, and I've played around with that setting using the above mentioned Pdo hack as necessary.

However, there is a standard way of iterating db results sets that all modern drivers, rdbms, etc support (including php+mysql) which is what I'm looking for:

1) issue large query
2) read one or several row at a time
3) when no more rows, all done

Only one (or several) rows are ever loaded into memory on the app side (and, if the db is smart enough, on the rdbms side, but that is irrelevant), never the entire result set.  (We can nitpick which settings specifically allow for this, but the point is that this is the desired economy of resources)

But, before each() even gets to 1), it is doing chunk()s extra step 0) splitting large query into smaller OFFSET queries, which is causing the problems (on the rdbms side, not app side)
